### PR TITLE
Recompile files when dependencies change

### LINKED ISSUE #3: Karma not watching preprocessed files?
When using this plugin and trying to watch my test files with karma, only changing the entry file causes the test files to be recompiled.

I have tried explicitly watching the rest of the test files but even though doing this causes the tests to run again, the test file is not recompiled! 😅

<details>
<summary>My Karma Config</summary>

```js
// Karma configuration
// Generated on Mon Dec 14 2015 17:29:26 GMT+0000 (GMT)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],


    // list of files / patterns to load in the browser
    files: [
      'test/client-test.js',
      {pattern: 'test/*', watched: true, included: false, served: false, nocache: false}
    ],


    // list of files to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'test/client-test.js': ['rollup']
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Firefox'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false,

    // Concurrency level
    // how many browser should be started simultanous
    concurrency: Infinity
  })
}
```
</details>

### PR DESCRIPTION
**Pull request by brianmhunt**
Closes #3

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -1,6 +1,21 @@
...
-var rollup = require('rollup');
+const fs = require('fs');
+const rollup = require('rollup');
+const debounce = require('debounce');
+
+const dependencies = new Map()
+const changedParents = new Set()
+const WAIT = 25
+
+const touchParents = debounce(() => {
```
Comment (TrySound) [inline lib/index.js]: node0.12 still should be supported.

Comment (brianmhunt) [inline lib/index.js]: What does that entail?

Comment (TrySound) [inline lib/index.js]: Don't use const, arrow functions, Map or Set without polyfill or transpilling. It's not so hard to write this stuff in es5.

Comment (brianmhunt) [inline lib/index.js]: I used the top-rated answer here http://stackoverflow.com/questions/28388885 as a guide.

Unless I misunderstood, it looks like for-of, Map/Set, and const are supported in Node 0.12.  So I removed the arrow functions and destructuring assignment.

ES6 is preferable not for ease, but for beauty. :smile:

**Comment by TrySound** [comment]
Here's more full table
http://kangax.github.io/compat-table/es6/
Also would be great to add some test cases.

**Comment by brianmhunt** [comment]
I also came across https://node.green

Agree on tests, but that'd be outside the scope of what I'm probably able to contribute for this issue. :smile:

**Comment by Kflash** [comment]
@brianmhunt The author seems inactive so I used a few bytes of your PR here and put together a better preprocessor. Located here: https://github.com/Kflash/rollup-plugin-karma

**Comment by jlmakes** [comment]
Thanks for this @brianmhunt, it has been working pretty well for me locally.

The other repository seems to have used your contribution initially, but abandoned them over the months and is now broken. I'm maintaining this repo now, and I'm going to get these changes merged shortly.

File: lib/index.js [hunk 2]
```diff
@@ -25,6 +42,29 @@ function createPreprocessor (args, config, logger, helper)
...
+                    // relies on this file, then trigger a recompilation of
+                    // it.
+                    for (var entry of dependencies.entries()) {
+                      var parent = entry[0];
+                      var dependList = entry[1]
+                      if (dependList.includes(file.originalPath)) {
+                        log.debug(" \n%s depends on \n\t%s\n    Recompiling it.",
+                          parent, file.originalPath);
+                        changedParents.add(parent);
+                        touchParents();
```
Comment (jlmakes) [inline lib/index.js]: cheeky function name...

File: lib/index.js [hunk 3]
```diff
@@ -25,6 +42,29 @@ function createPreprocessor (args, config, logger, helper)
                 .rollup(rollupConfig)
                 .then(function (bundle)
                 {
+                    // Map this file to the dependencies that Rollup just
+                    // compiled.
+                    dependencies.set(
+                      file.originalPath,
+                      bundle.modules
+                        .map(function (b) { return b.id })
```
Comment (jlmakes) [inline lib/index.js]: Counter intuitive that `id` is the actually the module file path

**Pull request closed without merging**